HPCC-14854 Add Roxie implementation of parallel PROJECT and PARSE

richardkchapman opened the pull request.

richardkchapman committed:
HPCC-14854 Add Roxie implementation of parallel PROJECT and PARSE

Add initial work wiring up threads/strands/junctions in Roxie. Remove all
usage of queryStream. Refactor setInput to include tracking of the
corresponding index, so that second pass can set up streams and insert
junctions as needed.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-14854 Add Roxie implementation of parallel PROJECT and PARSE

Refactoring onCreate vs constructor calls to ensure ctx is available sooner,
so that we can pass RowManager to the junction creation code.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

ghalliday committed:
HPCC-14825 Helper classes for managing multiple strands

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

richardkchapman committed:
HPCC-14854 Add Roxie implementation of parallel PROJECT and PARSE

Add code to support parallel project.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-14854 Add Roxie implementation of parallel PROJECT and PARSE

Use a unique allocator per strand.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-14854 Add Roxie implementation of parallel PROJECT and PARSE

Support hint to force order preservation, and refactor a little ready for
supporting more activities.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-14854 Add Roxie implementation of parallel PROJECT and PARSE

Refactor rather more to make it easier to support more stranded activities.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-14854 Add Roxie implementation of parallel PROJECT and PARSE

Fix remaining issues picked up by regression suite run with strands
forced on.

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman commented:
@jakesmith Please review.

Note that the full Roxie regression suite has passed with "forceNumStrands=4" which flushed out most of the issues I hope. I have reviewed Gavin's bits and he has reviewed mine, but a third pair of eyes would not hurt.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on roxie/ccd/ccdactivities.cpp:
```diff
@@ -5164,10 +5165,12 @@ class CRoxieRemoteActivity : public CRoxieSlaveActivity
             remoteGraph->beforeExecute();
             Owned<IFinalRoxieInput> input = remoteGraph->startOutput(0, remoteExtractBuilder.size(), remoteExtractBuilder.getbytes(), false);
-            IEngineRowStream &stream = input->queryStream();
+            Owned <IStrandJunction> junction;
```
trivial: space after Owned

jakesmith commented:
ccdserver.cpp too big for github to show diffs, so will list line #'s and comment here

trivial : ccdserver (1541)
> // Typically set from hints, common to many stranded activities

Not to all?

jakesmith commented:
trivial : ccdserver.cpp (1559)
> // Could consider some similar option for numStrands...

Appears to be left over from previous commit, before numStrands option was added

jakesmith commented:
question : ccdserver.cpp (1749) - CRoxieServerStrandedLateStartActivity

, doesn't look like it's used. What's it intended for?

jakesmith commented:
question: ccdserver.cpp (1359)
> if (!connected)

Is it okay to connectDependencies() the once? Could the input / streams change, e.g. if conditionals involved ?

jakesmith commented:
trivial: ccdserver.cpp(6202)

extra blank line

jakesmith commented:
minor: ccdserver.cpp (19977)
>        : StrandProcessor(_parent, _inputStream, true), helper(_helper)

needsAllocator=true ...

I thought it would need an allocator, but don't see where ParseProcessor uses it.

jakesmith commented:
style : ccdserver.cpp (19977)
>        : StrandProcessor(_parent, _inputStream, true), helper(_helper)

indented elsewhere.

jakesmith commented:
@richardkchapman - some minor comments, but looks good.

Now need to start slotting this into Thor.
I wonder how much effort it would be to get base activities similar enough (between Thor and Roxie), so that e.g. StrandProcessor could be common to both engines.

richardkchapman commented:
A lot, I suspect. Getting the input interface common was hard enough!

I'm not sure more commoning would make sense at the moment. Your ThorDataLink is a little different anyway

richardkchapman commented:
StrandedLateStart is not used yet but will be used for a stranded filter, for example.

richardkchapman committed:
HPCC-14854 Add Roxie implementation of parallel PROJECT and PARSE

Minor updates following review.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
> Is it okay to connectDependencies() the once? Could the input / streams change, e.g. if conditionals involved ?
Yes - conditionals retain the links to the true and false streams and select between them.

> needsAllocator=true ...
> I thought it would need an allocator, but don't see where ParseProcessor uses it.
Well spotted.  It creates one in the helper classes - and doesn't use this one.  I think it should be addressed in a different PR (I'll open a JIRA HPCC-14876), because it isn't a trivial issue.

jakesmith commented:
So this PR is ready to merge I think.
@richardkchapman

richardkchapman committed:
HPCC-14854 Add Roxie implementation of parallel PROJECT and PARSE

Regression suite detected an issue when NOT running stranded, related to IF
stopping branches that had not been started.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK
